### PR TITLE
#1145 Date formatters

### LINKED ISSUE #1145: Date formatters -- tidy up
As per [this comment thread](https://github.com/openmsupply/openmsupply-client/pull/1138#discussion_r840111999)

The current `useFormatDate` seems to be returning US-style dates -- we need something that can localise the date display regardless of where it is used.

Also -- I'm not sure why `useFormatDate` needs to be a "hook" -- it doesn't actually use any React state or other hooks, it just returns a plain JS function.

My suggestion would be to remove `useFormatDate` and consolidate date formatters into the existing "Formatters" module, specifically taking advantage of `date-fns` locale-aware functionality, although not sure the best file/function structure -- let's discuss!

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fix #1145

Large number of file changes, but most are just changing existing implementations to use the new localised hook `useFormatDateTime` (in DateUtils).

Deprecated `useFormatDate`

Quite a bit of fiddling around for not much actual change in the end, but at least the `useFormatDateTime` hook will  be easy to add additional localised date/time formatters to going forward.

Was planning to just make utility functions and not create a hook, but we need to access `useCurrentLanguage` so needs to be a hook.

For now, it just users browser language to determine whether to use US or GB for English, but I've made a new issue to investigate this further: #1150

### CODE REVIEW

File: packages/common/src/ui/layout/tables/columns/types.ts [hunk 1]
```diff
@@ -39,7 +39,7 @@ export type ColumnDataAccessor<T extends RecordWithId> = (params: {
 
 export type Translators = {
   t: ReturnType<typeof useTranslation>;
-  d: ReturnType<typeof useFormatDate>;
+  d: (date: string | number | Date) => string;
```
Comment (CarlosNZ) [inline packages/common/src/ui/layout/tables/columns/types.ts:42]: Could be used for several different date formatting methods (though only 2 currently), but all should follow this signature

File: packages/common/src/utils/formatters/formatters.ts [hunk 2]
```diff
@@ -6,9 +6,6 @@ export const Formatter = {
   // TODO: Do something better than naively rounding to 2 decimal places
   tax: (tax: number, withParens = true) =>
     `${withParens ? '(' : ''}${(tax ?? 0).toFixed(2)}%${withParens ? ')' : ''}`,
-  date: (date: Date): string =>
-    `${date.getDate()}/${date.getMonth() + 1}/${date.getFullYear()}`,
-  localisedDate: (date: Date): string => format(date, 'P'),
```
Comment (CarlosNZ) [inline packages/common/src/utils/formatters/formatters.ts:11]: Never used, can't see a reason for it.

File: packages/requisitions/src/RequestRequisition/DetailView/RequestLineEdit/ItemCharts/ConsumptionHistory.tsx [hunk 3]
```diff
@@ -29,13 +29,9 @@ export const ConsumptionHistory: React.FC<ConsumptionHistoryProps> = ({
...
   const theme = useTheme();
-  const d = useFormatDate();
+  const { customDate } = useFormatDateTime();
   const { data, isLoading } = useRequestLineChartData(id);
-  const dateFormatter = (date: string) =>
-    d(new Date(date), {
-      month: 'short',
-      day: '2-digit',
-    });
+  const dateFormatter = (date: string) => customDate(date, 'dd MMM');
```
Comment (CarlosNZ) [inline packages/requisitions/src/RequestRequisition/DetailView/RequestLineEdit/ItemCharts/ConsumptionHistory.tsx:34]: Does the same thing as what it replaces, just uses `date-fns` functions (localised) rather that Intl.DateTime

File: packages/common/src/intl/utils/DateUtils.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  isToday,
+  isThisMonth,
+  isAfter,
+  isBefore,
+  isEqual,
+  format,
+  parseISO,
+  fromUnixTime,
+} from 'date-fns';
+import { enGB, enUS, fr, ar } from 'date-fns/locale';
```
Comment (CarlosNZ) [inline packages/common/src/intl/utils/DateUtils.ts:16]: Not thrilled about having to import these. Ideally we would lazy-load them as required, as part of the Intl Context (updating when "changeLanguage" is run), but not sure how to do this / if it's worth it.

File: packages/common/src/intl/utils/DateUtils.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        : enGB
+      : getLocaleObj[language];
+
+  const localisedDate = (date: Date | string | number): string =>
+    format(dateInputHandler(date), 'P', { locale });
+
+  const customDate = (
+    date: Date | string | number,
+    formatString: string
+  ): string => format(dateInputHandler(date), formatString, { locale });
```
Comment (CarlosNZ) [inline packages/common/src/intl/utils/DateUtils.ts:65]: Strictly we can do all date formatting with the one method (customDate), but useful to create a handful of "preset" formats (added as required)

File: packages/common/src/intl/utils/DateUtils.ts (hunk 4 above)
Comment (mark-prins) [reply packages/common/src/intl/utils/DateUtils.ts:16]: agree - not ideal. would be nice to lazy load if we can.. but that's a problem for another day!

File: packages/common/src/intl/utils/DateUtils.ts (hunk 5 above)
Comment (mark-prins) [reply packages/common/src/intl/utils/DateUtils.ts:65]: yes, the 'preset' idea is important. Good to have something like the expiry date formatted consistently throughout, and easy to update if required

File: packages/common/src/utils/formatters/formatters.ts (hunk 2 above)
Comment (mark-prins) [reply packages/common/src/utils/formatters/formatters.ts:11]: I think a remnant from a time when the api was returning dates, or being parsed on fetch

File: packages/requisitions/src/RequestRequisition/DetailView/RequestLineEdit/ItemCharts/ConsumptionHistory.tsx (hunk 3 above)
Comment (mark-prins) [reply packages/requisitions/src/RequestRequisition/DetailView/RequestLineEdit/ItemCharts/ConsumptionHistory.tsx:34]: potentially a candidate for a preset date format too?

**Review by mark-prins — APPROVED** [review]
🎉  I like that a lot - very nicely wrapping up a few problems with formatting and inconsistent usage 😁

File: packages/requisitions/src/RequestRequisition/DetailView/RequestLineEdit/ItemCharts/ConsumptionHistory.tsx (hunk 3 above)
Comment (CarlosNZ) [reply packages/requisitions/src/RequestRequisition/DetailView/RequestLineEdit/ItemCharts/ConsumptionHistory.tsx]: Yup: 78394429

**Pull request merged**